src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp

```diff
@@ -1143,55 +1142,28 @@ void C2_MacroAssembler::string_equals(Register a1, Register a2,
     bne(tmp1, tmp2, DONE);
   } bgtz(cnt1, NEXT_WORD);
 
-  // Last longword.  In the case where length == 4 we compare the
-  // same longword twice, but that's still faster than another
-  // conditional branch.
-  // cnt1 could be 0, -1, -2, -3, -4 for chars; -4 only happens when
-  // length == 4.
-  add(tmp1, a1, cnt1);
-  ld(tmp1, Address(tmp1, 0));
-  add(tmp2, a2, cnt1);
-  ld(tmp2, Address(tmp2, 0));
-  bne(tmp1, tmp2, DONE);
-  j(SAME);
-
-  bind(SHORT);
-  Label TAIL03, TAIL01;
-
-  // 0-7 bytes left.
-  test_bit(t0, cnt1, 2);
-  beqz(t0, TAIL03);
-  {
-    lwu(tmp1, Address(a1, 0));
-    add(a1, a1, 4);
-    lwu(tmp2, Address(a2, 0));
-    add(a2, a2, 4);
-    bne(tmp1, tmp2, DONE);
-  }
-
-  bind(TAIL03);
-  // 0-3 bytes left.
-  test_bit(t0, cnt1, 1);
-  beqz(t0, TAIL01);
-  {
-    lhu(tmp1, Address(a1, 0));
-    add(a1, a1, 2);
-    lhu(tmp2, Address(a2, 0));
-    add(a2, a2, 2);
+  if (!AvoidUnalignedAccesses) {
+    // Last longword.  In the case where length == 4 we compare the
+    // same longword twice, but that's still faster than another
+    // conditional branch.
+    // cnt1 could be 0, -1, -2, -3, -4 for chars; -4 only happens when
+    // length == 4.
+    add(tmp1, a1, cnt1);
+    ld(tmp1, Address(tmp1, 0));
+    add(tmp2, a2, cnt1);
+    ld(tmp2, Address(tmp2, 0));
     bne(tmp1, tmp2, DONE);
+    j(SAME);
   }
 
-  bind(TAIL01);
-  if (elem_size == 1) { // Only needed when comparing 1-byte elements
-    // 0-1 bytes left.
-    test_bit(t0, cnt1, 0);
-    beqz(t0, SAME);
-    {
-      lbu(tmp1, Address(a1, 0));
-      lbu(tmp2, Address(a2, 0));
-      bne(tmp1, tmp2, DONE);
-    }
-  }
+  bind(SHORT);
+  ld(tmp1, Address(a1));
+  ld(tmp2, Address(a2));
+  xorr(tmp1, tmp1, tmp2);
+  neg(cnt1, cnt1);
+  slli(cnt1, cnt1, LogBitsPerByte);
+  sll(tmp1, tmp1, cnt1);
+  bnez(tmp1, DONE);
```

Comment:
that is hard to say.

OoO arches such as thead - don't care about the location of xor opcode here

In order uarches, such as u74/hifive might be affected by such change. however, the memory at address a1/a2 very likely would already be in the l1d cache, due to previous accesses in the same function, so it will be pretty cheap.
u74 is dual-issue, so it may execute these two loads (from l1d$) in parallel, and will wait for them to finish anyway, no matter what is next opcode.

To say for sure, need to check with jmh test org.openjdk.bench.java.lang.StringEquals on hifive

Reply:
Before the PR

```
Benchmark                      Mode  Cnt     Score   Error  Units
StringEquals.almostEqual       avgt   25  1214.131 ± 4.400  ns/op
StringEquals.almostEqualUTF16  avgt   25  1213.310 ± 7.156  ns/op
StringEquals.different         avgt   25    20.102 ± 2.306  ns/op
StringEquals.differentCoders   avgt   25    14.780 ± 1.147  ns/op
StringEquals.equal             avgt   25  1218.393 ± 5.275  ns/op
StringEquals.equalsUTF16       avgt   25  1216.750 ± 4.383  ns/op
```


With this PR

```
Benchmark                      Mode  Cnt   Score   Error  Units
StringEquals.almostEqual       avgt   25  28.584 ± 1.178  ns/op
StringEquals.almostEqualUTF16  avgt   25  28.375 ± 1.052  ns/op
StringEquals.different         avgt   25  19.572 ± 1.031  ns/op
StringEquals.differentCoders   avgt   25  14.969 ± 2.348  ns/op
StringEquals.equal             avgt   25  28.603 ± 0.148  ns/op
StringEquals.equalsUTF16       avgt   25  29.217 ± 1.969  ns/op
```

Xor moved

```
Benchmark                      Mode  Cnt   Score   Error  Units
StringEquals.almostEqual       avgt   25  28.455 ± 1.068  ns/op
StringEquals.almostEqualUTF16  avgt   25  28.244 ± 0.920  ns/op
StringEquals.different         avgt   25  18.940 ± 0.831  ns/op
StringEquals.differentCoders   avgt   25  14.566 ± 1.298  ns/op
StringEquals.equal             avgt   25  27.891 ± 0.606  ns/op
StringEquals.equalsUTF16       avgt   25  28.294 ± 0.913  ns/op
```

hard to say

Reply:
Thanks for trying this out. Another issue here is that 8-byte memory accesses at address a1/a2 would exceed the range for strings whose size is smaller than wordSize. I am not quite sure whether that is safe to do. Could we just incorporate changes which only resolves the unaligned access problem here?

Reply:
On an fpga I can see these numbers with perfnorm profiler:

Just this PR:

Secondary result "org.openjdk.bench.java.lang.StringEquals.equal:IPC":
  1.245 ±(99.9%) 0.037 insns/clk [Average]
  (min, avg, max) = (1.234, 1.245, 1.258), stdev = 0.010
  CI (99.9%): [1.209, 1.282] (assumes normal distribution)

This PR + moving xor:

Secondary result "org.openjdk.bench.java.lang.StringEquals.equal:IPC":
  1.239 ±(99.9%) 0.050 insns/clk [Average]
  (min, avg, max) = (1.224, 1.239, 1.253), stdev = 0.013
  CI (99.9%): [1.190, 1.289] (assumes normal distribution)

Reply:
You are right, if the object is located at the end of an allocated memory region ( and nothing past it) then this might produce sigsegv.

I'll ty to modify old version of string_equals to use TAIL logic for the tail of long strings, lwu+lhu+lbu is slower than ld but still faster than falling to misaligned access emulator.

Reply:
new version results in jmh:

```
Benchmark                      Mode  Cnt   Score   Error  Units
StringEquals.almostEqual       avgt   25  29.459 ± 0.173  ns/op
StringEquals.almostEqualUTF16  avgt   25  29.833 ± 0.596  ns/op
StringEquals.different         avgt   25  19.896 ± 1.024  ns/op
StringEquals.differentCoders   avgt   25  14.986 ± 1.748  ns/op
StringEquals.equal             avgt   25  31.174 ± 0.226  ns/op
StringEquals.equalsUTF16       avgt   25  32.718 ± 0.982  ns/op
```

should not read past an object boundaries anymore